src/DynamoCore/Engine/EngineController.cs

```diff
@@ -518,8 +513,18 @@ private void OnLibraryLoaded()
         /// </summary>
         private void LibraryLoaded(object sender, LibraryServices.LibraryLoadedEventArgs e)
         {
-            if(e.LibraryPaths.Any())
+            if (e.LibraryPaths.Any())
+            {
+                // Need to make compiled custom nodes available before resetting the VM
+                OnRequestCustomNodeRegistration();
                 OnLibraryLoaded();
```

Comment:
So now it looks like the custom node definitions could be scheduled to be added to the VM before it is reset in `OnLibraryLoaded`, which could again wipe out all of the definitions potentially. Wouldn't it be safer to call `OnRequestCustomNodeRegistration` after `OnLibraryLoaded`?

Reply:
This is where the method names might be misleading. I named this `OnRequestCustomNodeRegistration` mainly because it calls  `RegisterCustomNodeDefinitionWithEngine` downstream. However, that function does not register anything with the `RunCore`, that's the misleading part. What this does is fill a queue that gets processed afterwards, when calling `OnLibraryLoaded`.

Reply:
I'm not sure I understand. The `pendingCustomNodeSyncData` queue is processed in `EngineController.ProcessPendingCustomNodeSyncData`, which is called in `HomeWorkspaceModel.Run`. `Run` is not called from `OnLibraryLoaded`. My question is, what is the harm in calling `OnRequestCustomNodeRegistration` after the VM is reset in `OnLibraryLoaded`?

Reply:
> Run is not called from OnLibraryLoaded

You are right, I had overlooked this. Looking at the sequence of calls, the most appropriate place for this code should be in `HomeWorkspaceModel.LibraryLoaded`, which is the event handler triggering the run. That way we avoid doing extra work if no home workspace is open. Do you agree?

Reply:
Please see my comments below.

Reply:
Please see my comments below.